compiler/rustc_mir_build/src/builder/matches/mod.rs

```diff
@@ -1484,6 +1488,66 @@ impl<'tcx> MatchTreeBranch<'tcx> {
     }
 }
 
+/// Collects the bindings for a [`MatchTreeSubBranch`], preserving the order they appear in the
+/// pattern, as though the or-alternatives chosen in this sub-branch were inlined.
+fn sub_branch_bindings<'tcx>(
+    parents: &[PatternExtraData<'tcx>],
+    leaf_bindings: &[SubpatternBindings<'tcx>],
+) -> Vec<Binding<'tcx>> {
+    // In the common case, all bindings will be in leaves. Allocate to fit the leaf's bindings.
+    let mut bindings = Vec::with_capacity(leaf_bindings.len());
+    let remainder = push_sub_branch_bindings(&mut bindings, Some(parents), leaf_bindings);
+    // Make sure we've included all bindings. We can end up with a non-`None` remainder if there's
+    // an unsimplifed or-pattern at the end that doesn't contain bindings.
+    if let Some(remainder) = remainder {
+        assert!(remainder.iter().all(|parent| parent.bindings.is_empty()));
+        assert!(leaf_bindings.is_empty());
+    }
+    bindings
+}
+
+/// Helper for [`sub_branch_bindings`]. Returns any bindings yet to be inlined.
+fn push_sub_branch_bindings<'c, 'tcx>(
+    flattened: &mut Vec<Binding<'tcx>>,
+    parents: Option<&'c [PatternExtraData<'tcx>]>,
+    leaf_bindings: &[SubpatternBindings<'tcx>],
+) -> Option<&'c [PatternExtraData<'tcx>]> {
+    match parents {
+        None => bug!("can't inline or-pattern bindings: already inlined all bindings"),
+        Some(mut parents) => {
```

Comment:
nit: you're starting the function with an `unwrap()`, I think it would be a bit clearer to not take an `Option` and unwrap `remainder` before the call below